operators/src/source/gdal_source/mod.rs

```diff
@@ -610,31 +615,55 @@ where
         (tile_x_size, tile_y_size),       // requested raster size
         None,                             // sampling mode
     )?;
-    let data_grid = Grid::new(tile_grid, buffer.data)?;
-    let no_data_value_grid = NoDataValueGrid::new(data_grid, no_data_value);
-    let grid_or_empty = GridOrEmpty::from(no_data_value_grid);
-    Ok(grid_or_empty)
+    let data_grid = Grid::new(tile_grid.clone(), buffer.data)?;
+
+    let dataset_mask_flags = get_mask_flags(dataset, raster_band_index as i32)?;
+
+    if dataset_mask_flags.is_all_valid() {
+        debug!("all pixels are valid --> skip no-data and mask handling.");
+        return Ok(MaskedGrid::new_with_data(data_grid).into());
+    }
+
+    if dataset_mask_flags.is_nodata() {
+        debug!("raster uses a no-data value --> use no-data handling.");
+        let no_data_value = dataset_params
+            .no_data_value
+            .or_else(|| rasterband.no_data_value())
+            .and_then(FromPrimitive::from_f64);
+        let no_data_value_grid = NoDataValueGrid::new(data_grid, no_data_value);
+        let grid_or_empty = GridOrEmpty::from(no_data_value_grid);
+        return Ok(grid_or_empty);
+    }
+
+    debug!("use mask based no-data handling.");
```

Comment:
maybe throw error when the mask flag indicates alpha channel?

Reply:
i think many tiffs use the alpha channel even if it is only true, false

Reply:
do you mean that there are a lot of tiffs where the mask flag is wrong? Maybe then we should add an override like for the nodata value and still throw an error if there is no override and the mask flag is alpha channel

Reply:
By override I meant, that, like for no data values, the user should be able to define a _concrete_ GdalMaskFlag for the dataset that overrides the maskflag that is stored in the tiff. Then, in case the tiff says `GMF_ALPHA` we can override it with `GMF_NODATA` and use the alpha values as a nodata mask. Sorry if my comment was unclear. I think this would be more flexible than your current solution.